sympy/calculus/util.py

```diff
@@ -112,3 +119,601 @@ def elm_domain(expr, intrvl):
+
+    There is, however, no consensus on interval division.
+
+    `X / Y = \{ z| \exists x \in X, y \in Y \mid y \neq 0, z = x/y\}`
+
+    Note: According to this definition the quotient of two Limits may not
+    be a Limits object but may be a union of Limits.
+
+    `X^n = \{ x^n \mid x \in X\}`
+
+    Note
+    ====
+
+    The main focus in the interval arithmetic is on the simplest way to calculate
+    upper and lower endpoints for the range of values of a function in one or more
+    variables. These barriers are not necessarily the supremum or infimum, since
+    the precise calculation of those values can be difficult or impossible.
+
+    Examples
+    ========
+
+    >>> from sympy import Limits, sin, exp, log, pi, E
+    >>> from sympy.abc import x
+
+    >>> Limits(0, 1) + Limits(1, 2)
+    <1, 3>
+
+    >>> Limits(0, 1) - Limits(0, 2)
+    <-2, 1>
+
+    >>> Limits(-2, 3)*Limits(-1, 1)
+    <-3, 3>
+
+    >>> Limits(1, 2)*Limits(3, 5)
+    <3, 10>
+
+    Note: `<a, b>^2` is not same as `<a, b>*<a, b>`
+
+    >>> Limits(-1, 1)**2
+    <0, 1>
+
+    >>> Limits(1, 3) < 4
+    True
+
+    >>> Limits(1, 3) < -1
+    False
+
+    Some elementary functions can also take intervals as input.
+    A function `f` evaluated for some interval `<a, b>` is defined as
+    `f(<a, b>) = \{ f(x) | a \le x \le b \}`
+
+    >>> sin(Limits(pi/6, pi/3))
+    <1/2, sqrt(3)/2>
+
+    >>> exp(Limits(0, 1))
+    <1, E>
+
+    >>> log(Limits(1, E))
+    <0, 1>
+
+    Some symbol in an experession can be substituted for an Interval.
+    But it doesn't necessarily evaluate the Interval for that expression.
+
+    >>> (x**2 + 2*x + 1).subs(x, Limits(-1, 1))
+    <-1, 4>
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Interval_arithmetic
+
+    .. [2] http://fab.cba.mit.edu/classes/S62.12/docs/Hickey_interval.pdf
+
+    Notes
+    =====
+
+    Do not use ``Limits`` for floating point interval arithmetic calculations
+    use ``mpmath.iv`` instead.
+    """
+
+    is_real = True
+
+    def __new__(cls, min, max):
+
+        min = _sympify(min)
+        max = _sympify(max)
+
+        inftys = [S.Infinity, S.NegativeInfinity]
+        # Only allow real intervals (use symbols with 'is_real=True').
+        if not (min.is_real or min in inftys) \
+           or not (max.is_real or max in inftys):
+            raise ValueError("Only real intervals are supported")
+
+        # Make sure that the created interval will be valid.
+        if max.is_comparable and min.is_comparable:
+            if max < min:
+                raise ValueError("Lower limit should be smaller than upper limit")
+
+        if max == min:
+            return max
+
+        return Basic.__new__(cls, min, max)
+
+    # setting the operation priority
+    _op_priority = 11.0
+
+    def _intersect(self, other):
+        res = Interval(self.min, self.max).intersect(
+            Interval(other.min, other.max))
+        return Limits(res.min, res.max)
```

Comment:
I suspect that ``res`` may not always be an ``Interval``. But if it is, then it has ``start`` and ``end``.